action_server/src/action_server/actions/place.py

```diff
@@ -21,122 +19,163 @@ class Place(Action):
+        # We assume we already got the object if previous action passed an arm, an object and this object has the
+        # required type, or the required type is a reference.
+        got_object_in_task = (
+            self.context.arm_designator is not None and (self.context.object_type == self.semantics.object.type or
+                                                         self.semantics.object.type == 'reference'))
+
+        object_in_gripper = False
+        if not got_object_in_task:
+            object_to_arm_dict = {arm.occupied_by.type: arm for arm_name, arm in self._robot.arms.iteritems() if
+                                  arm.occupied_by and arm.occupied_by.is_a(self.semantics.object.type)}
+
+            if bool(object_to_arm_dict):
+                object_in_gripper = True
+                arm = object_to_arm_dict[self.semantics.object.type]
+                self.context.arm_designator = \
+                    ArmDesignator({arm.side: arm})
+
+        got_object = got_object_in_task or object_in_gripper
+
+        # If precondition not met, request prior action from the task manager
+        if not got_object:
+            # Request pick_up action
+            self._config_result.required_context = {'action': 'pick-up',
+                                                    'object': config.semantics['object']}
+            if 'source-location' in config.semantics:
+                self._config_result.required_context['source-location'] = config.semantics['source-location']
+            return
+        # Now we can assume we picked up the item!
+
+        # We should have navigated to the place where we should hand over
+        at_destination = (self.context.location is not None and
+                          (self.context.location == self.semantics.target_location or
+                           self.semantics.target_location.type == 'reference'))
+
+        if not at_destination:
+            # Request navigation action
+            self._config_result.required_context = {'action': 'navigate-to',
+                                                    'target-location': config.semantics['target-location']}
+            return
+        # We can now assume that we are at the destination for handover!
 
         self._config_result.succeeded = True
+        return
 
     def _start(self):
-        # Resolve the arm we want to place with
-        if self._arm_designator:
-            arm = self._arm_designator.resolve()
-            if arm:
-                self._arm = arm
-
-        # Which item do we want to place? The object in the hand we indicated
-        item_to_place = robot_smach_states.util.designators.Designator(self._arm.occupied_by, resolve_type=Entity)
-
-        # In the case of a task like "Find the coke and place it on the table" grasping is implied, so grasp it here
-        if item_to_place.resolve():
-            self._object_designator = item_to_place
+        self._robot.speech.speak("Bring the action!")
```

Comment:
This is now called after the navigation part. We could do this at the end of the configure step to make sure the robot says it to the operator immediately.